src/fparser/two/tests/test_parser.py

```diff
@@ -0,0 +1,71 @@
+# Copyright (c) 2018 Science and Technology Facilities Council
+# All rights reserved.
+#
+# Modifications made as part of the fparser project are distributed
+# under the following license:
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met:
+#
+# 1. Redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright
+# notice, this list of conditions and the following disclaimer in the
+# documentation and/or other materials provided with the distribution.
+#
+# 3. Neither the name of the copyright holder nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+''' Module containing tests for the parser file '''
+
+import pytest
+from fparser.two.parser import ParserFactory
+from fparser.common.readfortran import FortranStringReader
+from fparser.two.Fortran2003 import NoMatchError
+
+
+@pytest.mark.xfail(reason="This works in isolation but fails with other "
+                   "tests. No idea why :-(")
```

Comment:
Wow, you're hitting lots of exciting problems in this Issue. I bet this is something to do with the fact that pytest probably imports all the modules it finds before running the tests. In other modules you do, e.g.:

    # this is required to setup the fortran2003 classes
    _ = ParserFactory().create(std="f2003")

Possibly this should be done as part of test set-up (c.f. a fixture?) rather than every time a module is imported? The fact that I can run e.g. `py.test test_comments.py test_parser.py` without trouble (presumably because test_comments sets-up the 2003 parser) is evidence for this.

Reply:
There were two problems. The first was that the setting up of class relationships in the subclass dictionary was not being reset before each ParserFactory().create() call which meant that the list of classes just got longer and longer each time we called create(). This caused an f2003 parser to accept fortran2008 after an f2008 parser had been instantiated. This is now fixed. The second problem is what the author suspects, but this only came to light in one f2008 test when I changed the last parser test to end with the creation of an f2003 parser. I've now changed all f2008 tests to  call ParserFactory().create(std="f2008"). Actually in practice only one test needed this to be added to make all failing tests to work but the fact it was this one looks random to me so changing all f2008 tests seemed the best option.